File: alpaca-tests/indi_telescope_monitor.cpp
/*
    INDI Telescope Monitor - Standalone test client
    
    This client connects to an INDI telescope driver and displays
    real-time status updates including position, state, tracking, etc.
    
    Copyright (C) 2026 Gord Tulloch
    
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.
*/

#include <iostream>
#include <iomanip>
#include <memory>
#include <cstring>
#include <cmath>
#include <unistd.h>
#include <signal.h>
#include <baseclient.h>
#include <basedevice.h>

class TelescopeMonitor : public INDI::BaseClient
{
public:
    TelescopeMonitor() : telescopeDevice(nullptr), connected(false), hasCoords(false) {}
    ~TelescopeMonitor() override = default;

    void setTelescopeName(const char *name) { telescopeName = name; }

protected:
    void newDevice(INDI::BaseDevice dp) override;
    void deleteDevice(INDI::BaseDevice dp) override;
    void newProperty(INDI::Property property) override;
    void updateProperty(INDI::Property property) override;
    void removeProperty(INDI::Property property) override {}
    void newMessage(INDI::BaseDevice dp, int messageID) override;
    void serverConnected() override;
    void serverDisconnected(int exit_code) override;

private:
    void displayStatus();
    void displayCoordinates();
    void displayState();
    void displayTracking();
    void displayPark();
    const char *getTelescopeStateString(IPState state);

    INDI::BaseDevice telescopeDevice;
    std::string telescopeName;
    bool connected;
    bool hasCoords;
    
    // Cached property values
    double currentRA = 0.0;
    double currentDEC = 0.0;
    double targetRA = 0.0;
    double targetDEC = 0.0;
    double currentAZ = 0.0;
    double currentALT = 0.0;
    std::string telescopeState = "UNKNOWN";
    bool isTracking = false;
    bool isParked = false;
    std::string trackMode = "UNKNOWN";
};

void TelescopeMonitor::newDevice(INDI::BaseDevice dp)
{
    std::string deviceName = dp.getDeviceName();
    
    if (deviceName == telescopeName)
    {
        std::cout << "Found telescope device: " << deviceName << std::endl;
        telescopeDevice = dp;
        
        // Connect to the device
        connectDevice(deviceName.c_str());
    }
}

void TelescopeMonitor::deleteDevice(INDI::BaseDevice dp)
{
    if (dp.getDeviceName() == telescopeName)
    {
        std::cout << "Telescope device removed: " << dp.getDeviceName() << std::endl;
        telescopeDevice = INDI::BaseDevice();
        connected = false;
    }
}

void TelescopeMonitor::newProperty(INDI::Property property)
{
    std::string deviceName = property.getDeviceName();
    std::string propertyName = property.getName();
    
    if (deviceName != telescopeName)
        return;
    
    // Watch for CONNECTION property to know when device is connected
    if (propertyName == "CONNECTION")
    {
        auto connectionSP = property.getSwitch();
        if (connectionSP)
        {
            auto connectSW = connectionSP->findWidgetByName("CONNECT");
            if (connectSW && connectSW->getState() == ISS_ON)
            {
                connected = true;
                std::cout << "\n=== TELESCOPE CONNECTED ===" << std::endl;
                displayStatus();
            }
        }
    }
}

void TelescopeMonitor::updateProperty(INDI::Property property)
{
    std::string deviceName = property.getDeviceName();
    std::string propertyName = property.getName();
    
    if (deviceName != telescopeName)
        return;
    
    bool shouldDisplay = false;
    
    // Equatorial coordinates (RA/Dec)
    if (propertyName == "EQUATORIAL_EOD_COORD")
    {
        auto coordsNP = property.getNumber();
        if (coordsNP)
        {
            auto raNP = coordsNP->findWidgetByName("RA");
            auto decNP = coordsNP->findWidgetByName("DEC");
            
            if (raNP && decNP)
            {
                currentRA = raNP->getValue();
                currentDEC = decNP->getValue();
                hasCoords = true;
                shouldDisplay = true;
            }
        }
    }
    
    // Target coordinates
    if (propertyName == "TARGET_EOD_COORD")
    {
        auto targetNP = property.getNumber();
        if (targetNP)
        {
            auto raNP = targetNP->findWidgetByName("RA");
            auto decNP = targetNP->findWidgetByName("DEC");
            
            if (raNP && decNP)
            {
                targetRA = raNP->getValue();
                targetDEC = decNP->getValue();
                shouldDisplay = true;
            }
        }
    }
    
    // Horizontal coordinates (Alt/Az)
    if (propertyName == "HORIZONTAL_COORD")
    {
        auto horizNP = property.getNumber();
        if (horizNP)
        {
            auto azNP = horizNP->findWidgetByName("AZ");
            auto altNP = horizNP->findWidgetByName("ALT");
            
            if (azNP && altNP)
            {
                currentAZ = azNP->getValue();
                currentALT = altNP->getValue();
                shouldDisplay = true;
            }
        }
    }
    
    // Telescope info (state)
    if (propertyName == "TELESCOPE_INFO")
    {
        shouldDisplay = true;
    }
    
    // Tracking state
    if (propertyName == "TELESCOPE_TRACK_STATE")
    {
        auto trackSP = property.getSwitch();
        if (trackSP)
        {
            auto trackOnSW = trackSP->findWidgetByName("TRACK_ON");
            if (trackOnSW)
            {
                isTracking = (trackOnSW->getState() == ISS_ON);
                shouldDisplay = true;
            }
        }
    }
    
    // Tracking mode
    if (propertyName == "TELESCOPE_TRACK_MODE")
    {
        auto trackModeSP = property.getSwitch();
        if (trackModeSP)
        {
            for (int i = 0; i < trackModeSP->count(); i++)
            {
                auto sw = trackModeSP->at(i);
                if (sw->getState() == ISS_ON)
                {
                    trackMode = sw->getLabel();
                    shouldDisplay = true;
                    break;
                }
            }
        }
    }
    
    // Park state
    if (propertyName == "TELESCOPE_PARK")
    {
        auto parkSP = property.getSwitch();
        if (parkSP)
        {
            auto parkedSW = parkSP->findWidgetByName("PARK");
            if (parkedSW)
            {
                isParked = (parkedSW->getState() == ISS_ON);
                shouldDisplay = true;
            }
        }
    }
    
    // Connection status
    if (propertyName == "CONNECTION")
    {
        auto connectionSP = property.getSwitch();
        if (connectionSP)
        {
            auto connectSW = connectionSP->findWidgetByName("CONNECT");
            auto disconnectSW = connectionSP->findWidgetByName("DISCONNECT");
            
            if (connectSW && connectSW->getState() == ISS_ON)
            {
                if (!connected)
                {
                    connected = true;
                    std::cout << "\n=== TELESCOPE CONNECTED ===" << std::endl;
                }
            }
            else if (disconnectSW && disconnectSW->getState() == ISS_ON)
            {
                connected = false;
                std::cout << "\n=== TELESCOPE DISCONNECTED ===" << std::endl;
            }
        }
    }
    
    // Display updated status
    if (connected && shouldDisplay)
    {
        displayStatus();
    }
}

void TelescopeMonitor::newMessage(INDI::BaseDevice dp, int messageID)
{
    if (dp.getDeviceName() == telescopeName)
    {
        std::string message = dp.messageQueue(messageID);
        std::cout << "[MSG] " << message << std::endl;
    }
}

void TelescopeMonitor::serverConnected()
{
    std::cout << "Connected to INDI server" << std::endl;
    std::cout << "Waiting for telescope device: " << telescopeName << std::endl;
}

void TelescopeMonitor::serverDisconnected(int exit_code)
{
    std::cout << "Disconnected from INDI server (exit code: " << exit_code << ")" << std::endl;
    connected = false;
}

const char *TelescopeMonitor::getTelescopeStateString(IPState state)
{
    switch (state)
    {
        case IPS_IDLE:
            return "IDLE";
        case IPS_OK:
            return "OK";
        case IPS_BUSY:
            return "BUSY";
        case IPS_ALERT:
            return "ALERT";
        default:
            return "UNKNOWN";
    }
}

void TelescopeMonitor::displayStatus()
{
    // Clear previous output (simple version - just add newlines)
    std::cout << "\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" << std::endl;
    std::cout << "  TELESCOPE STATUS MONITOR" << std::endl;
    std::cout << "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" << std::endl;
    
    displayCoordinates();
    displayState();
    displayTracking();
    displayPark();
    
    std::cout << "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" << std::endl;
}

void TelescopeMonitor::displayCoordinates()
{
    std::cout << std::fixed << std::setprecision(6);
    
    std::cout << "\n📍 COORDINATES:" << std::endl;
    
    if (hasCoords)
    {
        // Convert RA from hours to HMS format
        int raHours = static_cast<int>(currentRA);
        int raMinutes = static_cast<int>((currentRA - raHours) * 60);
        double raSeconds = ((currentRA - raHours) * 60 - raMinutes) * 60;
        
        // Convert Dec to DMS format
        int decSign = (currentDEC >= 0) ? 1 : -1;
        double absDec = std::abs(currentDEC);
        int decDegrees = static_cast<int>(absDec);
        int decMinutes = static_cast<int>((absDec - decDegrees) * 60);
        double decSeconds = ((absDec - decDegrees) * 60 - decMinutes) * 60;
        
        std::cout << "  Current Position:" << std::endl;
        std::cout << "    RA  = " << std::setw(2) << std::setfill('0') << raHours << "h "
                  << std::setw(2) << raMinutes << "m "
                  << std::setw(5) << std::setprecision(2) << raSeconds << "s"
                  << "  (" << std::setprecision(6) << currentRA << " hours = "
                  << (currentRA * 15.0) << "°)" << std::endl;
        
        std::cout << "    Dec = " << (decSign >= 0 ? "+" : "-")
                  << std::setw(2) << std::setfill('0') << decDegrees << "° "
                  << std::setw(2) << decMinutes << "' "
                  << std::setw(5) << std::setprecision(2) << decSeconds << "\""
                  << "  (" << std::setprecision(6) << (decSign * absDec) << "°)" << std::endl;
        
        std::cout << "    Alt = " << std::setprecision(2) << currentALT << "°" << std::endl;
        std::cout << "    Az  = " << std::setprecision(2) << currentAZ << "°" << std::endl;
    }
    else
    {
        std::cout << "  No coordinate data available yet" << std::endl;
    }
}

void TelescopeMonitor::displayState()
{
    std::cout << "\n⚙️  STATE:" << std::endl;
    
    // Get the EQUATORIAL_EOD_COORD property state to determine telescope state
    auto device = telescopeDevice;
    if (device.isValid())
    {
        auto coordsProp = device.getProperty("EQUATORIAL_EOD_COORD");
        if (coordsProp.isValid())
        {
            IPState state = coordsProp.getState();
            const char *stateStr = getTelescopeStateString(state);
            
            std::string status;
            if (isParked)
                status = "PARKED";
            else if (state == IPS_BUSY)
                status = "SLEWING";
            else if (isTracking)
                status = "TRACKING";
            else if (state == IPS_IDLE)
                status = "IDLE";
            else if (state == IPS_OK)
                status = "OK";
            else
                status = stateStr;
            
            std::cout << "  Status: " << status << std::endl;
        }
    }
}

void TelescopeMonitor::displayTracking()
{
    std::cout << "\n🎯 TRACKING:" << std::endl;
    std::cout << "  Enabled: " << (isTracking ? "YES" : "NO") << std::endl;
    std::cout << "  Mode:    " << trackMode << std::endl;
}

void TelescopeMonitor::displayPark()
{
    std::cout << "\n🅿️  PARK:" << std::endl;
    std::cout << "  Parked:  " << (isParked ? "YES" : "NO") << std::endl;
}

// Global pointer for signal handler
TelescopeMonitor *g_monitor = nullptr;

void signal_handler(int signum)
{
    std::cout << "\n\nReceived signal " << signum << ", shutting down..." << std::endl;
    if (g_monitor)
        g_monitor->disconnectServer();
    exit(signum);
}

void usage(const char *program)
{
    std::cout << "Usage: " << program << " [options]" << std::endl;
    std::cout << "Options:" << std::endl;
    std::cout << "  -h HOST      INDI server host (default: localhost)" << std::endl;
    std::cout << "  -p PORT      INDI server port (default: 7624)" << std::endl;
    std::cout << "  -d DEVICE    Telescope device name (default: alpaca_telescope)" << std::endl;
    std::cout << "  -?           Show this help" << std::endl;
}

int main(int argc, char *argv[])
{
    std::string host = "localhost";
    int port = 7624;
    std::string deviceName = "alpaca_telescope";
    
    // Parse command line arguments
    int opt;
    while ((opt = getopt(argc, argv, "h:p:d:?")) != -1)
    {
        switch (opt)
        {
            case 'h':
                host = optarg;
                break;
            case 'p':
                port = std::atoi(optarg);
                break;
            case 'd':
                deviceName = optarg;
                break;
            case '?':
            default:
                usage(argv[0]);
                return 1;
        }
    }
    
    std::cout << "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" << std::endl;
    std::cout << "  INDI Telescope Monitor" << std::endl;
    std::cout << "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" << std::endl;
    std::cout << "Connecting to: " << host << ":" << port << std::endl;
    std::cout << "Telescope:     " << deviceName << std::endl;
    std::cout << "Press Ctrl+C to exit" << std::endl;
    std::cout << "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n" << std::endl;
    
    // Create telescope monitor
    TelescopeMonitor monitor;
    g_monitor = &monitor;
    
    // Set up signal handler for clean exit
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    // Set telescope name
    monitor.setTelescopeName(deviceName.c_str());
    
    // Set server address
    monitor.setServer(host.c_str(), port);
    
    // Connect to server
    if (!monitor.connectServer())
    {
        std::cerr << "Failed to connect to INDI server at " << host << ":" << port << std::endl;
        return 1;
    }
    
    // Main loop - just wait for updates
    while (true)
    {
        sleep(1);
    }
    
    return 0;
}

File: alpaca-tests/indi_telescope_monitor.py
#!/usr/bin/env python3
"""
INDI Telescope Monitor - Simple Python test client

This client connects to an INDI telescope driver and displays
real-time status updates including position, state, tracking, etc.

Usage:
    python3 indi_telescope_monitor.py [--host HOST] [--port PORT] [--device DEVICE]
"""

import sys
import time
import socket
import xml.etree.ElementTree as ET
from datetime import datetime
import argparse

class INDITelescopeMonitor:
    def __init__(self, host='localhost', port=7624, device='alpaca_telescope'):
        self.host = host
        self.port = port
        self.device_name = device
        self.sock = None
        self.connected = False
        self.device_connected = False
        
        # Telescope state
        self.current_ra = 0.0
        self.current_dec = 0.0
        self.target_ra = 0.0
        self.target_dec = 0.0
        self.current_alt = 0.0
        self.current_az = 0.0
        self.is_tracking = False
        self.is_parked = False
        self.track_mode = "UNKNOWN"
        self.telescope_state = "UNKNOWN"
        self.has_coords = False
        
        self.buffer = b''
        
    def connect(self):
        """Connect to INDI server"""
        try:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.connect((self.host, self.port))
            self.sock.settimeout(0.1)  # Non-blocking with timeout
            self.connected = True
            print(f"Connected to INDI server at {self.host}:{self.port}")
            
            # Request properties from our device
            self.send_message(f'<getProperties version="1.7" device="{self.device_name}"/>\n')
            
            return True
        except Exception as e:
            print(f"Failed to connect: {e}")
            return False
    
    def disconnect(self):
        """Disconnect from INDI server"""
        if self.sock:
            self.sock.close()
            self.sock = None
        self.connected = False
        print("Disconnected from INDI server")
    
    def send_message(self, message):
        """Send XML message to INDI server"""
        if self.sock:
            self.sock.sendall(message.encode('utf-8'))
    
    def receive_data(self):
        """Receive and process data from INDI server"""
        try:
            data = self.sock.recv(4096)
            if not data:
                return False
            
            self.buffer += data
            
            # Process complete XML messages
            while True:
                # Find complete XML element
                start = self.buffer.find(b'<')
                if start == -1:
                    self.buffer = b''
                    break
                
                # Look for various closing tags
                end_tags = [b'</setNumberVector>', b'</setSwitchVector>', 
                           b'</setTextVector>', b'</setBLOBVector>',
                           b'</defNumberVector>', b'</defSwitchVector>',
                           b'</defTextVector>', b'</defBLOBVector>',
                           b'</message>', b'</delProperty>']
                
                end = -1
                for tag in end_tags:
                    pos = self.buffer.find(tag, start)
                    if pos != -1:
                        if end == -1 or pos < end:
                            end = pos + len(tag)
                
                if end == -1:
                    # No complete message yet
                    break
                
                # Extract and process message
                message = self.buffer[start:end]
                self.buffer = self.buffer[end:]
                
                try:
                    self.process_message(message.decode('utf-8'))
                except Exception as e:
                    pass  # Ignore parsing errors
            
            return True
            
        except socket.timeout:
            return True  # Timeout is OK, just means no data
        except Exception as e:
            print(f"Error receiving data: {e}")
            return False
    
    def process_message(self, xml_str):
        """Process incoming XML message"""
        try:
            root = ET.fromstring(xml_str)
            
            # Check device name
            device = root.get('device', '')
            if device != self.device_name:
                return
            
            prop_name = root.get('name', '')
            state = root.get('state', '')
            
            # Connection status
            if prop_name == 'CONNECTION':
                for elem in root.findall('.//oneSwitch'):
                    if elem.get('name') == 'CONNECT' and elem.text == 'On':
                        if not self.device_connected:
                            self.device_connected = True
                            print("\n=== TELESCOPE CONNECTED ===")
                            self.display_status()
            
            # Equatorial coordinates
            elif prop_name == 'EQUATORIAL_EOD_COORD':
                for elem in root.findall('.//oneNumber'):
                    name = elem.get('name')
                    if name == 'RA':
                        self.current_ra = float(elem.text)
                        self.has_coords = True
                    elif name == 'DEC':
                        self.current_dec = float(elem.text)
                
                self.telescope_state = state
                self.display_status()
            
            # Target coordinates
            elif prop_name == 'TARGET_EOD_COORD':
                for elem in root.findall('.//oneNumber'):
                    name = elem.get('name')
                    if name == 'RA':
                        self.target_ra = float(elem.text)
                    elif name == 'DEC':
                        self.target_dec = float(elem.text)
            
            # Horizontal coordinates
            elif prop_name == 'HORIZONTAL_COORD':
                for elem in root.findall('.//oneNumber'):
                    name = elem.get('name')
                    if name == 'AZ':
                        self.current_az = float(elem.text)
                    elif name == 'ALT':
                        self.current_alt = float(elem.text)
                self.display_status()
            
            # Tracking state
            elif prop_name == 'TELESCOPE_TRACK_STATE':
                for elem in root.findall('.//oneSwitch'):
                    if elem.get('name') == 'TRACK_ON' and elem.text == 'On':
                        self.is_tracking = True
                    elif elem.get('name') == 'TRACK_OFF' and elem.text == 'On':
                        self.is_tracking = False
                self.display_status()
            
            # Tracking mode
            elif prop_name == 'TELESCOPE_TRACK_MODE':
                for elem in root.findall('.//oneSwitch'):
                    if elem.text == 'On':
                        self.track_mode = elem.get('name', 'UNKNOWN')
                self.display_status()
            
            # Park state
            elif prop_name == 'TELESCOPE_PARK':
                for elem in root.findall('.//oneSwitch'):
                    if elem.get('name') == 'PARK' and elem.text == 'On':
                        self.is_parked = True
                    elif elem.get('name') == 'UNPARK' and elem.text == 'On':
                        self.is_parked = False
                self.display_status()
            
            # Messages
            elif root.tag == 'message':
                message = root.get('message', '')
                timestamp = root.get('timestamp', datetime.now().isoformat())
                print(f"[MSG {timestamp}] {message}")
                
        except ET.ParseError:
            pass  # Ignore incomplete XML
        except Exception as e:
            pass  # Ignore other errors
    
    def display_status(self):
        """Display current telescope status"""
        if not self.device_connected:
            return
        
        print("\n" + "="*50)
        print("  TELESCOPE STATUS MONITOR")
        print("="*50)
        
        # Coordinates
        print("\n📍 COORDINATES:")
        if self.has_coords:
            # Convert RA from hours to HMS
            ra_hours = int(self.current_ra)
            ra_minutes = int((self.current_ra - ra_hours) * 60)
            ra_seconds = ((self.current_ra - ra_hours) * 60 - ra_minutes) * 60
            
            # Convert Dec to DMS
            dec_sign = '+' if self.current_dec >= 0 else '-'
            abs_dec = abs(self.current_dec)
            dec_degrees = int(abs_dec)
            dec_minutes = int((abs_dec - dec_degrees) * 60)
            dec_seconds = ((abs_dec - dec_degrees) * 60 - dec_minutes) * 60
            
            print(f"  Current Position:")
            print(f"    RA  = {ra_hours:02d}h {ra_minutes:02d}m {ra_seconds:05.2f}s  "
                  f"({self.current_ra:.6f} hours = {self.current_ra * 15.0:.6f}°)")
            print(f"    Dec = {dec_sign}{dec_degrees:02d}° {dec_minutes:02d}' {dec_seconds:05.2f}\"  "
                  f"({self.current_dec:.6f}°)")
            print(f"    Alt = {self.current_alt:.2f}°")
            print(f"    Az  = {self.current_az:.2f}°")
        else:
            print("  No coordinate data available yet")
        
        # State
        print("\n⚙️  STATE:")
        status_map = {
            'Idle': 'IDLE',
            'Ok': 'OK', 
            'Busy': 'SLEWING',
            'Alert': 'ALERT'
        }
        
        if self.is_parked:
            status = "PARKED"
        elif self.telescope_state == 'Busy':
            status = "SLEWING"
        elif self.is_tracking:
            status = "TRACKING"
        else:
            status = status_map.get(self.telescope_state, self.telescope_state)
        
        print(f"  Status: {status}")
        
        # Tracking
        print("\n🎯 TRACKING:")
        print(f"  Enabled: {'YES' if self.is_tracking else 'NO'}")
        print(f"  Mode:    {self.track_mode}")
        
        # Park
        print("\n🅿️  PARK:")
        print(f"  Parked:  {'YES' if self.is_parked else 'NO'}")
        
        print("="*50)
    
    def run(self):
        """Main loop"""
        if not self.connect():
            return
        
        print(f"Monitoring telescope: {self.device_name}")
        print("Press Ctrl+C to exit\n")
        
        try:
            while True:
                if not self.receive_data():
                    print("Connection lost")
                    break
                time.sleep(0.1)
        except KeyboardInterrupt:
            print("\n\nShutting down...")
        finally:
            self.disconnect()

def main():
    parser = argparse.ArgumentParser(
        description='INDI Telescope Monitor - Display real-time telescope status'
    )
    parser.add_argument('--host', default='localhost',
                      help='INDI server host (default: localhost)')
    parser.add_argument('--port', type=int, default=7624,
                      help='INDI server port (default: 7624)')
    parser.add_argument('--device', default='alpaca_telescope',
                      help='Telescope device name (default: alpaca_telescope)')
    
    args = parser.parse_args()
    
    print("="*50)
    print("  INDI Telescope Monitor")
    print("="*50)
    print(f"Connecting to: {args.host}:{args.port}")
    print(f"Telescope:     {args.device}")
    print("Press Ctrl+C to exit")
    print("="*50 + "\n")
    
    monitor = INDITelescopeMonitor(args.host, args.port, args.device)
    monitor.run()

if __name__ == '__main__':
    main()
